### PR TITLE
[mind2] Utilize sine type breathing function

### PR DESCRIPTION
**Pull request by spiiroin**

### CODE REVIEW

File: sysfs-led-main.c [hunk 1]
```diff
@@ -547,17 +550,31 @@ sysfs_led_generate_ramp_half_sin(int ms_on, int ms_off)
   int steps_on  = (n * ms_on + t / 2) / t;
   int steps_off = n - steps_on;
 
-  const float m_pi_2 = (float)M_PI_2;
-
+  /* Calclulate a non-zero value for each step on the ramp.
```
Comment (mlehtima) [inline sysfs-led-main.c:553]: Typo: Calclulate -> Calculate

File: sysfs-led-main.c [hunk 2]
```diff
@@ -638,12 +711,24 @@ static void
 sysfs_led_generate_ramp(int ms_on, int ms_off)
 {
   switch( led_control_breath_type(&led_control) ) {
+  case LED_RAMP_SINE:
+    sysfs_led_generate_ramp_sine(ms_on, ms_off, false);
+    break;
+
+  case LED_RAMP_SAWTOOTH:
+      sysfs_led_generate_ramp_sawtooth(ms_on, ms_off);
```
Comment (mlehtima) [inline sysfs-led-main.c:719]: This and the next case have indentation mismatch to other cases.

File: sysfs-led-main.h [hunk 3]
```diff
@@ -41,6 +42,15 @@ typedef enum {
 
   /** Step function used for emulating blinking via sw breathing */
   LED_RAMP_HARD_STEP = 2,
+
+  /* Triangle waveform */
```
Comment (mlehtima) [inline sysfs-led-main.h:46]: This and next entry have only one * when other start with two

File: sysfs-led-util.h [hunk 4]
```diff
@@ -26,6 +27,53 @@
 
 # include <stdbool.h>
 
+/* ========================================================================= *
+ * UTILITY
+ * ========================================================================= */
+
+/** Clamp float value to given range
+ */
```
Comment (mlehtima) [inline sysfs-led-util.h:35]: This could be just single line comment like those below

File: sysfs-led-main.c [hunk 5]
```diff
@@ -623,6 +640,62 @@ sysfs_led_generate_ramp_hard_step(int ms_on, int ms_off)
...
+
+  /* Set a non-zero intensity value for each step on the ramp.
+   */
+  int i = 0;
+
+  for( ; i < steps_on; ++i )
+    sysfs_led_breathe.value[i] = led_util_trans(i, 0, steps_on, 1, 255);;
+
+  for( ; i < steps_tot; ++i )
+      sysfs_led_breathe.value[i] = led_util_trans(steps_tot - i, 0, steps_off, 1, 255);
```
Comment (mlehtima) [inline sysfs-led-main.c:682]: Indentation mismatch.

File: sysfs-led-util.c [hunk 6]
```diff
@@ -107,13 +108,16 @@ led_util_open_file(int *fd_ptr, const char *path)
...
 int
 led_util_scale_value(int in, int max)
 {
-  int out = (in * max + 128) / 255;
-
-  return (out < 0) ? 0 : (out < max) ? out : max;
+  int out = 0;
+  if( in > 0 )
+      out = led_util_trans(in, 1, 255, 1, max);
+    return out;
```
Comment (mlehtima) [inline sysfs-led-util.c:120]: Indentation mismatch.

File: sysfs-led-main.c [hunk 7]
```diff
@@ -623,6 +640,62 @@ sysfs_led_generate_ramp_hard_step(int ms_on, int ms_off)
...
+  mce_log(LL_DEBUG, "delay=%d, steps_on=%d, steps_off=%d",
+          sysfs_led_breathe.delay, steps_on, steps_off);
+}
+
+/** Generate sawtooth wave for use from breathing timer
+ */
+static void
+sysfs_led_generate_ramp_sawtooth(int ms_on, int ms_off)
+{
+    sysfs_led_generate_ramp_triangle(ms_on + ms_off, 0);
```
Comment (mlehtima) [inline sysfs-led-main.c:696]: Indentation in rest of the file is 2 spaces not 4

**Review by mlehtima — CHANGES_REQUESTED** [review]

File: sysfs-led-main.c (hunk 1 above)
Comment (spiiroin) [reply sysfs-led-main.c:553]: fixed

File: sysfs-led-main.c (hunk 2 above)
Comment (spiiroin) [reply sysfs-led-main.c]: fixed

File: sysfs-led-main.h (hunk 3 above)
Comment (spiiroin) [reply sysfs-led-main.h]: yup

File: sysfs-led-util.h (hunk 4 above)
Comment (spiiroin) [reply sysfs-led-util.h]: I guess the idea was to use this form but then something happened - lets go with the shorter version as you suggested.

File: sysfs-led-util.c (hunk 6 above)
Comment (spiiroin) [reply sysfs-led-util.c]: fixed

File: sysfs-led-main.c (hunk 7 above)
Comment (spiiroin) [reply sysfs-led-main.c]: fixed

File: sysfs-led-main.c (hunk 5 above)
Comment (spiiroin) [reply sysfs-led-main.c]: fixed

**Review by mlehtima — APPROVED** [review]

**Pull request merged**